oauth2client/contrib/multiprocess_file_storage.py

```diff
@@ -0,0 +1,261 @@
+
+This module provides the MultiprocessFileStorage class that:
+* Is tied to a single credential via a user-specified key. This key can be used
+  to distinguish between multiple users, client ids, and/or scopes.
+* Can be safely accessed and refreshed across threads and processes.
+
+Process & thread safety guarantees the following behavior:
+* If one process refreshes a credential, subsequent refreshes from other
+  processes will re-fetch the credentials from the file instead of performing
+  an http request.
+* If two processes attempt to refresh concurrently, only one process will be
+  able to acquire the lock and refresh, with the deadlock caveat below.
+* The interprocess lock will not deadlock, instead, the if a process can not
+  acquire the interprocess lock within INTERPROCESS_LOCK_DEADLINE it will
+  allow refreshing the credential but will not write the updated credential to
+  disk, This logic happens during every lock cycle - if the credentials are
+  refreshed again it will retry locking and writing as normal.
+
+"""
+
+import base64
+import json
+import logging
+import os
+import threading
+
+import fasteners
+from six import iteritems
+
+from oauth2client.client import Credentials
+from oauth2client.client import Storage as BaseStorage
+
+INTERPROCESS_LOCK_DEADLINE = 1
+logger = logging.getLogger(__name__)
+_backends = {}
+_backends_lock = threading.Lock()
+
+
+def _get_backend(filename):
+    """A helper method to get or create a backend with thread locking.
+
+    There should only be one backend per-file per-process."""
+
+    # This method prevents race conditions.
+    try:
+        return _backends[filename]
+    except KeyError:
+        pass
+
+    with _backends_lock:
+        backend = _MultiprocessStorageBackend(filename)
+        _backends[filename] = backend
+
+    return backend
+
+
+class _MultiprocessStorageBackend(object):
+
+    def __init__(self, filename):
+        self._file = None
+        self._filename = filename
+        self._process_lock = fasteners.InterProcessLock(
+            '{}.lock'.format(filename))
+        self._thread_lock = threading.Lock()
+        self._read_only = False
+        self._credentials = {}
+
+    def _create_file_if_needed(self):
+        """Creates the an empty credential storage file if it does not
+        exist."""
+        if self._read_only:
+            return False
+
+        if not os.path.exists(self._filename):
+            open(self._filename, 'a+b').close()
+            logging.info('Credential file {} created'.format(self._filename))
+            return False
+
+        return True
+
+    def _load_credentials(self):
+        """(Re-)loads the credentials from the file."""
+        if not self._create_file_if_needed():
+            return
+
+        self._credentials.update(self._load_credentials_file(self._file))
+
+        logger.debug('Read credential file')
+
+    def _load_credentials_file(self, fh):
+        credentials = {}
```

Comment:
"Effective Java", Second Edition, Item 45: 
"Minimize the scope of local variables": this local variable isn't actually mutated until line 131 of the current draft. In the places where it appears prior to that, you could eliminate its use by just using the literal value `{}`. I think you should introduce this local variable just above line 127 of the current draft.